### PR TITLE
Fix some potential memory leaks

### LINKED ISSUE #8655: [BUG] There are some potential device memory leaks in `AbstractGpuCoalesceIterator`
[In the following code](https://github.com/NVIDIA/spark-rapids/blob/branch-23.08/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala#L419-L427) in class `AbstractGpuCoalesceIterator`, the `cbFromIter` will be leaked on the `if` branch, because `GpuFilter` will return a new batch without closing the input one.

```
      val cbFromIter = iter.next()

      var cb = if (inputFilterExpression.isDefined) {
        // If we have reached the cuDF limit once, proactively filter batches
        // after that first limit is reached.
        GpuFilter(cbFromIter, inputFilterExpression.get)
      } else {
        cbFromIter
      }
```

Similar issues at [line 451 and 454](https://github.com/NVIDIA/spark-rapids/blob/branch-23.08/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala#L451-L454).
```
                val filteredDown = GpuFilter(concatAllAndPutOnGPU(), filterExpression)
                closeOnExcept(filteredDown) { _ =>
                  // filter the incoming batch as well
                  closeOnExcept(GpuFilter(cb, filterExpression)) { filteredCb =>
```

### PR DESCRIPTION
**Pull request by firestarman**
fixes #8655

This PR is to fix some potential memory leaks in GPU coalescing operator.

### CODE REVIEW

**Comment by firestarman** [comment]
build

**Comment by abellina** [comment]
Nice find. The bug I think is happening because `GpuFilter.apply` used to call `close` the incoming batch but it doesn't anymore: https://github.com/NVIDIA/spark-rapids/pull/7864/files#diff-7718c93b5e463c82611c693aa760c934343e3d8047dd800f1d0c00e9ab8afa52L396

A new method was added "filterAndClose" and I think we may want to do that instead. From me messing about with the code it looks like the only external consumer to this overload of `GpuFilter.apply` is the coalesce class where you found the leak, so can we make this apply private and then use `filterAndClose`?? it will also retry if all the projections are deterministic, and I think that's the intended behavior.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala [hunk 1]
```diff
@@ -448,11 +448,18 @@ abstract class AbstractGpuCoalesceIterator(
...
-                val filteredDown = GpuFilter(concatAllAndPutOnGPU(), filterExpression)
+                val filteredDown = withResource(concatAllAndPutOnGPU()) { concatCB =>
+                  GpuFilter(concatCB, filterExpression)
+                }
                 closeOnExcept(filteredDown) { _ =>
-                  // filter the incoming batch as well
-                  closeOnExcept(GpuFilter(cb, filterExpression)) { filteredCb =>
-                    cb = null // null out `cb` to prevent multiple close calls
+                  // filter the incoming batch if not be filtered.
+                  val filteredCb = if (inputFilterExpression.isEmpty) {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:456]: Why filter only if the filter isEmpty?

**Review by revans2 — COMMENTED** [review]
That for fixing this. It looks like I caused this when I introduced filterAndClose to GpuFilter and changed some semantics.  Would it be better to just change the calls to GpuFilter.apply to GpuFilter.filterAndClose ?

**Comment by abellina** [comment]
BTW I don't think this bug is triggered in 23.04 (when the retry change happened) or 23.06 because the re-hash work made it so we discard the coalesce goal object if the logic in `GpuShuffledHashJoinExec.prepareBuildBatchesForJoin` decides we need a single batch.

Because:

if `needSingleBuildBatch` is true, `sizeOverflow` is false (unless we are in tests, then other cases arise), so we always call `getBuildBatchOptimizedAndClose`, which disregards the filter. 

If `needSingleBuildBatch` is false (`sizeOverflow` must be true, unless we are in tests), then we never call `getAsSingleBatch` which takes into account the `buildGoal`.

So, if I am seeing this right, as of 23.04+ we are not filtering at all. Not until: https://github.com/NVIDIA/spark-rapids/pull/8659 which now takes into account `buildGoal` when `needSingleBuildBatch` is false.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala (hunk 1 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:456]: When the filter expression is defined, the batch will be filtered at line 424.
```
      var cb = if (inputFilterExpression.isDefined) {
        // If we have reached the cuDF limit once, proactively filter batches
        // after that first limit is reached.
        withResource(cbFromIter)(GpuFilter(_, inputFilterExpression.get))
      } else {
        cbFromIter
      }
```

**Comment by firestarman** [comment]
build

**Review by revans2 — APPROVED** [review]

**Review by abellina — APPROVED** [review]

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
@abellina @revans2  Could you review it again ?
I have fixed some logic defects for the case when there are 2 or more input batches after entering the filtering mode, and added some unit tests for it.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Review by revans2 — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala [hunk 2]
```diff
@@ -447,72 +457,67 @@ abstract class AbstractGpuCoalesceIterator(
...
-                  numBytes = filteredBytes
+                    } // end of "while(filteredCbIter.hasNext)"
+                    // 3) Setup the filter
+                    inputFilterTier = Some(filterTier)
+                    logWarning("Switched to null-filtering mode. This coalesce iterator " +
+                      "succeeded to fit rows under the cuDF limit only after null " +
+                      "filtering. Please try increasing your partition count.")
+                    numRows = filteredNumRows
+                    numBytes = filteredBytes
+                  } else {
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala:513]: nice catch

**Review by abellina — APPROVED** [review]

**Pull request merged**